src/Uno.UI/UI/Xaml/UIElement.cs

```diff
@@ -1043,6 +1043,12 @@ internal virtual void PrepareForRecycle()
 		{
 			ClearPointerStateOnRecycle();
 
+
+			// It's important to clear the index on the way out (recycle) and not wait to set it on the way in (reuse)
+			// because this property is used e.g. in ItemsControl.ContainerFromIndexInner which is used by e.g.
+			// ListView to get a clicked container.
+			ClearValue(ItemsControl.IndexForItemContainerProperty);
```

Comment:
Very nice find! We have similar properties if I remember correctly, could you take a look and see if it needs to be done for others?

Reply:
This is scary in UIElement. I think `ClearValue` should throw if the actual instance type isn't the same as (or a derived from) the property owner type.

Better to override this in `ItemsControl`. In ItemsControl, there is `CleanUpContainer` method that I'm not sure whether it should be called as well or not. @jeromelaban Is `CleanUpContainer` what you're referring to?

Reply:
@Youssef1313  I'm pretty sure that's not true. We set this on ListViewItem anyway and not on ListView itself (also, it's not throwing for me). CleanUpContainer is called conditionally (I'm guessing for performance reasons) where it more thoroughly cleans up the container (e.g. remove the content). We need to clear IndexForItemContainerProperty even if we don't CleanUpContainer. This is the case for the other properties, but we use IndexForItemContainerProperty to get ContainerFromIndex, otherwise we could leave it stale like the other properties until it's reused (where it will get set to a new value anyway).

Reply:
Oh that property is attached, so indeed there is no concern with throwing.

> We need to clear IndexForItemContainerProperty even if we don't CleanUpContainer

I didn't mean to clear `IndexForItemContainerProperty` only in `CleanupContainer`. The question is about whether we need to call CleanupContainer in `PrepareForRecycle`

Reply:
> Better to override this in `ItemsControl`. In ItemsControl, there is `CleanUpContainer` method that I'm not sure whether it should be called as well or not. @jeromelaban Is `CleanUpContainer` what you're referring to?

Indeed, I misread the location of the change. This should be in `ItemsControl.CleanupContainer` or `ClearContainerForItemOverride`:

https://github.com/unoplatform/uno/blob/66b8c63dc66119c334a7b9a676ffbd9f7a7c9ff6/src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs#L1105

not in `UIElement`. We may call this conditionally (probably incorrectly), but it may be platform dependent. Which platforms do you see this issue happening?

Reply:
It seems like the conditional calling is indeed incorrect, as unconditionally calling `CleanUpContainer` also seems to fix a problem with scrolling that I noticed a couple of times where scrolling all the way down would clip the last 2 items. In that case, putting the `ClearValue` call inside `CleanUpContainer` makes more sense. Ideally, I would prefer if we knew exactly when we could get away with not calling `CleanUpContainer`, but due to the inconsistency of things, I guess we can just clean the container every time until further notice.

The issue seems to appear mainly on WASM and skia.

Reply:
Could you point to the conditional call you're seeing?

Reply:
`CleanUpContainer` is called if `clearContainer` is true in `RecycleViewForItem` (I've changed it in the last couple of commits, so you'll find it in the before side of the diff)